prog/prog.go

```diff
@@ -212,46 +288,52 @@ func (p *Prog) insertBefore(c *Call, calls []*Call) {
 }
 
 // replaceArg replaces arg with arg1 in call c in program p, and inserts calls before arg call.
-func (p *Prog) replaceArg(c *Call, arg, arg1 *Arg, calls []*Call) {
-	if arg.Kind != ArgConst && arg.Kind != ArgResult && arg.Kind != ArgPointer && arg.Kind != ArgUnion {
-		panic(fmt.Sprintf("replaceArg: bad arg kind %v", arg.Kind))
-	}
-	if arg1.Kind != ArgConst && arg1.Kind != ArgResult && arg1.Kind != ArgPointer && arg.Kind != ArgUnion {
-		panic(fmt.Sprintf("replaceArg: bad arg1 kind %v", arg1.Kind))
-	}
-	if arg.Kind == ArgResult {
-		delete(arg.Res.Uses, arg)
-	}
+func (p *Prog) replaceArg(c *Call, arg, arg1 Arg, calls []*Call) {
 	for _, c := range calls {
 		sanitizeCall(c)
 	}
 	p.insertBefore(c, calls)
-	// Somewhat hacky, but safe and preserves references to arg.
-	uses := arg.Uses
-	*arg = *arg1
-	arg.Uses = uses
-	if arg.Kind == ArgResult {
-		delete(arg.Res.Uses, arg1)
-		arg.Res.Uses[arg] = true
+	switch a := arg.(type) {
+	case *ConstArg:
+		*a = *arg1.(*ConstArg)
+	case *ResultArg:
+		if a.Res != nil {
+			delete(*a.Res.(ArgUsed).Used(), arg)
+		}
+		used := *arg.(ArgUsed).Used()
+		*a = *arg1.(*ResultArg)
+		*arg.(ArgUsed).Used() = used
+		if a.Res != nil {
+			delete(*a.Res.(ArgUsed).Used(), arg1)
```

Comment:
Turns out we do. I added comments to clarify what's going on.
The `if a.Res != nil` block in the beginning and in the end refer to different `a`, do we can't combine them.